python/mujinasync/asynctcp.py

```diff
@@ -334,37 +338,66 @@ def SpinOnce(self, timeout=0):
                 timeout = 0 # force no wait at select later since we have a new connection to report right away
 
         # pool all the sockets
-        socketConnections = {}
+        socketConnections: dict[socket.socket, tuple[TcpServerClient, TcpConnection]] = {}
         for serverClient in self._servers + self._clients:
             for connection in serverClient._connections:
+                if connection.connectionSocket is None:
+                    continue
                 if connection.receiveBuffer.size >= connection.receiveBuffer.capacity:
                     connection.receiveBuffer.capacity *= 2
                 rsockets.append(connection.connectionSocket)
                 if connection.sendBuffer.size > 0:
                     wsockets.append(connection.connectionSocket)
-                xsockets.append(connection.connectionSocket)
                 socketConnections[connection.connectionSocket] = (serverClient, connection)
 
-        # select
-        while True:
-            try:
-                rlist, wlist, xlist = select.select(rsockets, wsockets, xsockets, timeout)
-                break
-            except (OSError, select.error) as e:
-                if e.args[0] != errno.EINTR:
-                    raise
+        with selectors.DefaultSelector() as selector:
+            for sock in rsockets:
+                try:
+                    selector.register(sock, selectors.EVENT_READ, data=sock)
+                except (OSError, ValueError) as e:
+                    log.warning('failed to register read socket %s: %s', sock, e)
+            
+            for sock in wsockets:
+                try:
+                    # check if already registered for read, then modify
+                    try:
+                        selector.modify(sock, selectors.EVENT_READ | selectors.EVENT_WRITE, data=sock)
+                    except KeyError:
+                        selector.register(sock, selectors.EVENT_WRITE, data=sock)
+                except (OSError, ValueError) as e:
+                    log.warning('failed to register write socket %s: %s', sock, e)
+            
+            # wait for events
+            while True:
+                try:
+                    events = selector.select(timeout if timeout > 0 else None)
+                    break
+                except (OSError, select.error) as e:
+                    if e.args[0] != errno.EINTR:
+                        raise
```

Comment:
Is it efficient to rebuild the selector each loop? Can we not maintain it and only change if the number of tracked fds changes?

Reply:
Updated, current method is:
- Create `self._selector` when `TcpContext` initialized. Close it when destoried.
- Register / modify sockets in selector's internal map while iterating all sockets. Since the original code is already iterating all sockets.
- Unregister all socket not in `self._servers` / `self._clients` from selector.
- Then handle socket IO, rest part is untouched.